store/file/header.go

```diff
@@ -0,0 +1,94 @@
+package file
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/celestiaorg/celestia-node/share"
+)
+
+const headerSize = 64
+
+type header struct {
+	version  fileVersion
+	fileType fileType
+
+	// Taken directly from EDS
+	shareSize  uint16
+	squareSize uint16
+
+	datahash share.DataHash
+}
+
+type fileVersion uint8
+
+const (
+	fileV0 fileVersion = iota
+)
+
+type fileType uint8
+
+const (
+	ods fileType = iota
+	q1q4
+)
+
+func (h *header) WriteTo(w io.Writer) (int64, error) {
+	b := bytes.NewBuffer(make([]byte, 0, headerSize))
+	_ = b.WriteByte(byte(h.version))
+	_ = b.WriteByte(byte(h.fileType))
+	_ = binary.Write(b, binary.LittleEndian, h.shareSize)
+	_ = binary.Write(b, binary.LittleEndian, h.squareSize)
+	_, _ = b.Write(h.datahash)
+	// write padding
+	_, _ = b.Write(make([]byte, headerSize-b.Len()-1))
+	return writeLenEncoded(w, b.Bytes())
+}
+
+func readHeader(r io.Reader) (*header, error) {
+	bytesHeader, err := readLenEncoded(r)
+	if err != nil {
+		return nil, err
+	}
+	if len(bytesHeader) != headerSize-1 {
+		return nil, fmt.Errorf("readHeader: read %d bytes, expected %d", len(bytesHeader), headerSize)
+	}
+	h := &header{
+		version:    fileVersion(bytesHeader[0]),
+		fileType:   fileType(bytesHeader[1]),
+		shareSize:  binary.LittleEndian.Uint16(bytesHeader[2:4]),
+		squareSize: binary.LittleEndian.Uint16(bytesHeader[4:6]),
+		datahash:   make([]byte, 32),
+	}
+
+	copy(h.datahash, bytesHeader[6:6+32])
+	return h, err
+}
+
+func writeLenEncoded(w io.Writer, data []byte) (int64, error) {
+	_, err := w.Write([]byte{byte(len(data))})
+	if err != nil {
+		return 0, err
+	}
+	return io.Copy(w, bytes.NewBuffer(data))
```

Comment:
`w.Write(data)` ?

Reply:
I don't want to manage amount bytes written manually and prefer io.Copy to do the job

Reply:
In my example you don't manage how many bytes you're writing. If it will fail - it will.

Reply:
For example, if you write 100 bytes with w.Write(data), it can write first 10 bytes out of 100 and return `10, nil`. While it is not an error, you need yo manually write the rest to the writer. I don't want to do it manually.

Reply:
As per `io.Writer` documentation (which sadly doesn't give any guarantees on the interface implementation, oof) it should work as: `It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early.`

So, if it writer less than `len(p)` we should have an error otherwise everything should be written fully.

I don't see a good argument to continue this thread now, let's leave it as `io.Copy + bytes.Buffer`.

Reply:
If one doesn't trust `Writer` inteface to return a non-nil error when n is less than `len(data)`, why one trusts `WriterTo` interface to be any better?

`io.Copy` will recognize that `bytes.Buffer` implements `io.WriterTo` method and forward the call directly:
https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/io/io.go;l=411
https://cs.opensource.google/go/go/+/refs/tags/go1.22.4:src/bytes/buffer.go;l=258

Here is the WriterTo documentation:
https://pkg.go.dev/io#WriterTo
It doesn't give any contracts while io.Writer docs require this:
> Write must return a non-nil error if it returns n < len(p). Write must not modify the slice data, even temporarily. 

I would at least give it a second thought. The code right now is more fragile (it relies on something that is not guaranteed while avoiding something with a specified contract) plus it does more job (creates a new bytes buffer, for example) than necessary.

Reply:
`WriteTo` implementation of bytes.Buffer does give guarantee of full write, and in case when not full data has been written it will return io.ErrShortWrite. Nonetheless I agree, that in discussed snippet ordinary `Write` should be sufficient and will be simplier. I'll will update code accordingly